### PR TITLE
HPCC-29331 Centralized log trace

### PR DESCRIPTION
**Pull request by rpastrana**
- Creates centralized trace handling class

- Removes distributed Global/caller/local id references

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by rpastrana** [comment]
@ghalliday please see second commit, I encountered some build-time difficulties abstracting out the logtrace fields from  IContextLogger into a dedicated interface in order to provide a base implementation which could be used anywhere IContextLogger is implemented, but wanted to ensure this approach coincides w/ your vision before I spend more time on it. Thanks.

**Review by ghalliday — CHANGES_REQUESTED** [review]
@rpastrana I don't see how most of the changes in the second commit tie into the jira.

In particular why the introduction of LogTracer?  It doesn't seem to match any of the comments on the first commit.  It is now no longer a self-contained light-weight class, it is a template class that reimplements the functionality.

I think we need to talk through at our next meeting.

Comments on 1st commit from an email:

> •	I would move into a separate jtrace.hpp/cpp.  This functionality isn't exclusive to logging, and some of the functionality will have nothing to do with logging.  Cleaner to keep it separate.
> •	No need to derive from CInterface.  I suspect this will always be a member.
> •	Use jlib rather than esp conventions for members - don't prefix with m_
> •	Default constructor should not allocate an id.  Either delete it or use default.  You don't want ids being created accidentally.
> •	HTTP_HEADER_HPCC_GLOBAL_ID don't need to be in the header - they should be private (in the cpp).
> •	Long term I would hope all the set() functions could be private, with other more abstract functions public.  That will only be clear as the class is used though.

**Comment by rpastrana** [comment]
> @rpastrana I don't see how most of the changes in the second commit tie into the jira.
> 
> In particular why the introduction of LogTracer? It doesn't seem to match any of the comments on the first commit. It is now no longer a self-contained light-weight class, it is a template class that reimplements the functionality.
> 
> I think we need to talk through at our next meeting.
> 
> Comments on 1st commit from an email:
> 
> > •	I would move into a separate jtrace.hpp/cpp.  This functionality isn't exclusive to logging, and some of the functionality will have nothing to do with logging.  Cleaner to keep it separate.
> > •	No need to derive from CInterface.  I suspect this will always be a member.
> > •	Use jlib rather than esp conventions for members - don't prefix with m_
> > •	Default constructor should not allocate an id.  Either delete it or use default.  You don't want ids being created accidentally.
> > •	HTTP_HEADER_HPCC_GLOBAL_ID don't need to be in the header - they should be private (in the cpp).
> > •	Long term I would hope all the set() functions could be private, with other more abstract functions public.  That will only be clear as the class is used though.

@ghalliday sure, let's talk anytime. The above comments resulted in several re-implementations of the exact same logic. The purpose of the LogTracer was to provide a base implementation of the trace related portion of the Interface.

**Comment by rpastrana** [comment]
image

File: thorlcr/graph/thgraph.cpp [hunk 1]
```diff
@@ -2689,41 +2686,39 @@ class CThorContextLogger : implements IContextLogger, public CSimpleInterface
...
+        return logTrace.queryLocalId();
     }
     virtual const char *queryCallerId() const override
     {
-        return callerId.str();
+        return logTrace.queryCallerId();
     }
     virtual void setHttpIdHeaders(const char *global, const char *caller)
     {
         if (global && *global)
```
Comment (ghalliday) [inline thorlcr/graph/thgraph.cpp:2709]: minor: logTrace.setHttpIdHeaderNames()...  hopefully to be deleted soon.
Same in a few other places.

File: system/jlib/jtrace.hpp [hunk 2]
```diff
@@ -15,11 +15,47 @@
...
+    const char* queryGlobalIdHTTPHeaderName() const { return globalIdHTTPHeaderName.get(); }
+    const char* queryCallerIdHTTPHeaderName() const { return callerIdHTTPHeaderName.get(); }
+
+    void setHttpIdHeaderNames(const char *global, const char *caller)
+    {
+        setGlobalIdHTTPHeaderName(global);
+        setCallerIdHTTPHeaderName(caller);
+    }
+
+    void setGlobalIdHTTPHeaderName(const char* headerName) { if (!isEmptyString(headerName)) globalIdHTTPHeaderName.set(headerName);};
```
Comment (ghalliday) [inline system/jlib/jtrace.hpp:50]: Should be possible to make this private (and delete it if it is only called from setHttpIdHeaderNames)

**Review by ghalliday — COMMENTED** [review]
@rpastrana a few minor comments to clean up the code.

File: system/jlib/jlog.cpp [hunk 3]
```diff
@@ -2830,39 +2827,38 @@ class DummyLogCtx : implements IContextLogger
...
-    virtual const char *queryLocalId() const
+    virtual const char *queryLocalId() const override
     {
-        return localId.str();
+        return logTrace.queryLocalId();
     }
-    virtual void setHttpIdHeaders(const char *global, const char *caller)
+    virtual void setHttpIdHeaderNames(const char *global, const char *caller) override
     {
         if (global && *global)
```
Comment (ghalliday) [inline system/jlib/jlog.cpp:2850]: logTrace.setHttpIdHeaderNames(global, caller);

(see comment earlier).  You can then delete the functions setGlobalIdHTTPHeaderName/setCallerIdHTTPHeaderName

The reason is to minimize the public interface

**Review by ghalliday — COMMENTED** [review]
@rpastrana not sure if you saw my previous comments.

**Comment by ghalliday** [comment]
@rpastrana please rebase and probably squash at the same time.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**